src/components/ConfigEditor/AzureCredentialsConfig.ts

```diff
@@ -0,0 +1,207 @@
+
+  if (typeof jsonData.tenantId === 'string' || typeof jsonData.clientId === 'string') {
+    try {
+      const azureCloud = resolveLegacyCloudName(jsonData.azureCloud);
+
+      return {
+        authType: jsonData.onBehalfOf ? 'clientsecret-obo' : 'clientsecret',
+        azureCloud: azureCloud,
+        tenantId: typeof jsonData.tenantId === 'string' ? jsonData.tenantId : '',
+        clientId: typeof jsonData.clientId === 'string' ? jsonData.clientId : '',
+        clientSecret: options.secureJsonFields.clientSecret ? concealedLegacy : undefined,
+      };
+    } catch (e) {
+      if (e instanceof Error) {
+        console.error('Unable to restore legacy credentials: %s', e.message);
+      }
+      return undefined;
+    }
+  }
+
+  return undefined;
+}
+
+function resolveLegacyCloudName(cloudName: string): AzureCloud {
+  if (!cloudName) {
+    return AzureCloud.Public;
+  }
+
+  switch (cloudName) {
+    case 'azuremonitor':
+      return AzureCloud.Public;
+    case 'chinaazuremonitor':
+      return AzureCloud.China;
+    case 'govazuremonitor':
+      return AzureCloud.USGovernment;
+    default:
+      throw new Error(`Azure cloud '${cloudName}' not supported by Azure Data Explorer datasource.`);
+  }
+}
+
+export function updateCredentials(
+  options: DataSourceSettings<any, any>,
+  credentials: AzureCredentials
+): DataSourceSettings<any, any> {
+  // Cleanup legacy credentials
+  options = {
+    ...options,
+    jsonData: {
+      ...options.jsonData,
+      azureCloud: undefined,
+      tenantId: undefined,
+      clientId: undefined,
+      onBehalfOf: undefined,
+    },
+  };
+
+  // Apply updated credentials
+  switch (credentials.authType) {
+    case 'msi':
+      if (!config.azure.managedIdentityEnabled) {
+        throw new Error('Managed Identity authentication is not enabled in Grafana config.');
+      }
+      options = {
+        ...options,
+        jsonData: {
+          ...options.jsonData,
+          azureCredentials: {
+            authType: 'msi',
+          },
+        },
+      };
+      break;
+
+    case 'clientsecret':
+    case 'clientsecret-obo':
+      options = {
+        ...options,
+        jsonData: {
+          ...options.jsonData,
+          azureCredentials: {
+            authType: credentials.authType,
+            azureCloud: credentials.azureCloud || getDefaultAzureCloud(),
+            tenantId: credentials.tenantId,
+            clientId: credentials.clientId,
+          },
+        },
+        secureJsonData: {
+          ...options.secureJsonData,
+          azureClientSecret:
+            typeof credentials.clientSecret === 'string' && credentials.clientSecret.length > 0
+              ? credentials.clientSecret
+              : undefined,
+        },
+        secureJsonFields: {
+          ...options.secureJsonFields,
+          azureClientSecret: credentials.clientSecret === concealed,
+          clientSecret: credentials.clientSecret === concealedLegacy,
```

Comment:
what happens if you don't update the legacy way of setting this?

Reply:
Assume that existing credentials stored in legacy format, existing secret stored in the `clientSecret` field on backend.

User goes to the datasource configuration UI, changes only `clientId` and clicks Save. Secret wasn't changed by the user and expected to be kept intact.

Credentials can only be saved in the new format, so the updated credentials will be stored in the new format, `azureCredentials` object. The new format normally keeps secret in another field - `azureClientSecret`. Since user didn't update the secret, the existing secret should not be lost. So this code has extra handling of this legacy secret scenario. If secret stored in `azureClientSecret` then keep it in `azureClientSecret`, if secret already stored in `clientSecret` then continue to keep it there.

These tests are covering this scenario:
https://github.com/kostrse/azure-data-explorer-datasource/blob/f104f5ccd24da0cee568c92bcb56ad37f47a67ac/src/components/ConfigEditor/AzureCredentialsConfig.test.ts#L100